### ISSUE TITLE
VBScriptEngine.AddHostType memory/handle leak

### ISSUE THREAD

**Issue by winkxoe** [microsoft/ClearScript#135]
It looks like a memory/handle leak was added sometime between versions 5.5.4 and 5.5.5 when calling AddHostTiype from an instance of VBScriptEngine. I think it also affects the JScriptEngine, but not the V8ScriptEngine.

**Comment by ClearScriptLib** [comment]
HI @winkxoe,

Can you provide more information? How are you detecting the leak? Do you know what objects are leaked?

Thanks!

**Comment by winkxoe** [comment]
I have a application that spins off a new thread every few seconds and
runs a simple VB script using VBScriptEngine. After it's done running
the script it disposes the VBScriptEngine. It also uses AddHostType to
add a class that enables the script to log to the same log file as the
rest of the application. We were using version 5.4.6 for a long time
without any problems. I upgraded to 5.5.6 and the application would grow
to over 9 GB of memory over a 24 hour period and slow the entire server
down. Then I tried 5.6.0 and had the same problem.

 

Then I did an awful lot of testing to try to isolate the problem. I went
back an tried versions 5.5.3, 5.5.4, and 5.5.5. Versions 5.5.3 and 5.5.4
work fine. Right now we're running 5.5.4 without the application
growing. It was also clear from my testing that it would not grow if I
removed the AddHostType call for the logging class. I tested the
V8ScriptEngine in a similar situation with the same AddHostType call for
the logging class. The application didn't appear to grow with V8.

 

It's probably not the best practice to create and dispose of so many
instances of the engine. In this case it seems to work for what we need
it for.

 

I hope this helps! Thank you for supporting ClearScript! It's a
wonderful tool!

**Comment by ClearScriptLib** [comment]
Hi @winkxoe,

A couple of questions:

1. Which [`AddHostType`](https://microsoft.github.io/ClearScript/Reference/html/Overload_Microsoft_ClearScript_ScriptEngine_AddHostType.htm) overload are you using?
1. Can you reproduce the leak simply by adding the `AddHostType` call? Or does it only happen if you call methods on the exposed type from script code?

Thanks!

**Comment by winkxoe** [comment]
I'm using AddHostType(string itemName, SystemType type);

 

I didn't have to call any methods on the exposed type. I actually
reduced the script to simply setting a string as shown in the example
below and I saw the leak.

 

Dim nextfile

nextfile = "The very next file"

**Comment by ClearScriptLib** [comment]
Hi @winkxoe,

Unfortunately we haven't been able to reproduce your issue so far with ClearScript 5.6.0. In a test program we construct a `VBScriptEngine` instance, expose the `Console` type, and execute a small script.  After 100,000 iterations the program's private memory size is around 12 MB.

Can you create a minimal sample program that reproduces the issue?

Thanks!

**Comment by winkxoe** [comment]
Can you try this chunk of code? I noticed that if I removed either the
AddHostType or the engine.Script["nextfile"] the leak seemed to go away.
However both of them together and it grows fast.

 

 

            while (!stop)

            {

                string text =

                    "Dim nextfile\n" +

                    "nextfile = \"The very next file\"";

 

                using (var engine = new VBScriptEngine())

                {

                    engine.AddHostType("Logger", typeof(Console));

                    engine.Script["nextfile"] = "";

                    engine.Execute(@text);

                }

            }

**Comment by ClearScriptLib** [comment]
@winkxoe,

Thanks for posting the sample code! We've reproduced the issue and will investigate.

Cheers!

**Comment by ClearScriptLib** [comment]
Hi @winkxoe,

This looks like a bug in [COM Interop](https://en.wikipedia.org/wiki/COM_Interop). We'll have a workaround in the next ClearScript release. In the meantime, you might be able to bypass the issue by replacing this:

``` C#
engine.Script["nextfile"] = "";
```

with this:

``` C#
engine.Script.nextfile = "";
```

Good luck!

**Comment by winkxoe** [comment]
Great! Thank you very much!

**Comment by ClearScriptLib** [comment]
Fixed in [ClearScript 6.0](https://github.com/microsoft/ClearScript/releases/tag/6.0.0).
